src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,209 @@
+            // Example:
+            // In the following case, from the perspective of type 'Store', we should not see duplicate
+            // errors related to type 'Address'
+            // public class Store
+            // {
+            //    [Required]
+            //    public int Id { get; set; }
+            //    public Address Address { get; set; }
+            // }
+            // public class Employee
+            // {
+            //    [Required]
+            //    public int Id { get; set; }
+            //    public Address Address { get; set; }
+            // }
+            // public class Address
+            // {
+            //    [Required]
+            //    public string Line1 { get; set; }
+            //    [Required]
+            //    public int Zipcode { get; set; }
+            //    [Required]
+            //    public string State { get; set; }
+            // }
+            var rootNodeValidationErrors = new Dictionary<Type, List<string>>();
+
+            Validate(modelType, visitedTypes, rootNodeValidationErrors);
+
+            foreach (var validationError in rootNodeValidationErrors)
+            {
+                foreach (var validationErrorMessage in validationError.Value)
+                {
+                    // Add error message to model state as exception to avoid
+                    // disclosing details to end user as SerializableError sanitizes the
+                    // model state errors having exceptions with a generic message when sending
+                    // it to the client.
+                    modelStateDictionary.TryAddModelError(
+                        validationError.Key.FullName,
+                        new InvalidOperationException(validationErrorMessage));
+                }
+            }
+        }
+
+        private void Validate(
+            Type modelType,
+            HashSet<Type> visitedTypes,
+            Dictionary<Type, List<string>> errors)
+        {
+            // Need not handle Key-Value pair specially here (for example, when the model type is Dictionary<,>
+            // which implements IEnumerable<KeyValuePair<TKey, TValue>>) as the model type here would be 
+            // KeyValuePair<TKey, TValue> where TKey and TValue are public properties which would also 
+            // be probed for Required attribute validation.
+            if (modelType.IsGenericType())
+            {
+                var enumerableOfT = modelType.ExtractGenericInterface(typeof(IEnumerable<>));
+                if (enumerableOfT != null)
+                {
+                    modelType = enumerableOfT.GetGenericArguments()[0];
+                }
+            }
+
+            if (ExcludeTypeFromValidation(modelType))
+            {
+                return;
+            }
+
+            // Avoid infinite loop in case of self-referencing properties
+            if (!visitedTypes.Add(modelType))
```

Comment:
Here we walk through the tree of a 'type' but in case of DefaultObjectValidator the walk is over the object itself, is it?...in that case wondering if we can have a common logic...like in the case of current PR even if the deserialized object is 'null', we still walk through the 'type' tree and add the errors to model state...